src/main/java/com/google/research/bleth/simulator/StatisticsState.java

```diff
@@ -83,6 +90,29 @@ public void writeBeaconsObservedPercentStats() {
         datastore.put(entity);
     }
 
+    /**
+     * Create and write multiple datastore entities, represent statistics about the intervals of time each beacon
+     * has been observed by at least one observer, i.e. has been detected by the resolver.
+     * @throws StatisticsAlreadyExistException if the simulation's observed statistics are already exists in the database.
+     */
+    public void writeBeaconsObservedStats() {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        Table<String, String, Double> beaconsIdToObservedStats = readBeaconsObservedStats(simulationId);
+        if (!beaconsIdToObservedStats.isEmpty()) {
+            throw new StatisticsAlreadyExistException(simulationId);
+        }
+
+        Map<String, Map<String, Double>> idToObservedStats = beaconsObservedStats.rowMap();
```

Comment:
Why do you need this change in representation?
You could just use `beaconsObservedStats.rowKeySet()` in the loop, and `beaconsObservedStats.row(beaconId)` on line 111.